### ISSUE TITLE
$dbh->ping documentation possibly not clear

### ISSUE THREAD

**Issue by dr-kd** [bucardo/dbdpg#121]
This test demonstrates - assumes a local docker container running:

```perl
#!/usr/bin/env perl                                                                                                       
use DBI;
use Test::More;
sub dbh {
    DBI->connect('dbi:Pg:host=127.0.0.1;dbname=mydb', 'myuser', 'mypass',
                 {AutoCommit => 1, RaiseError => 1, PrintError => 0});
}

$dbh = dbh();
eval { $dbh->selectrow_array('select 1') };
ok ! $@, "1. got result after connection";
system 'docker restart myserver';
sleep 2;
eval { $dbh->ping };
ok $@, '2. $dbh->ping should raise exception if connection is stale, no?';
my @res = eval { $dbh->selectrow_array('select 1') };
ok ! @res, "3. query after failed ping should return no data";
ok $@, "4.  Query after failed ping should raise exception";
$dbh = dbh() if $@;
system 'docker restart myserver';
sleep 2;

# looks like we get no exception here either.                                                                             
my @res =  eval { $dbh->selectrow_array('select 1') };

ok ! @res, "5.  result should be empty due to stale connection";
ok $@, "4b. should have thrown an exception due to same";
$dbh = dbh() if $@;
@res = eval { $dbh->selectrow_array('select 1') };
ok @res && !$@, "6. we detected stale and reconnected";
done_testing;
```

[edited the test to try to ensure the logic is correct]

**Comment by turnstep** [comment]
ping is designed to always return something, not to throw an error. You can trust it to return 0 if the connection is bad. Also see the pg_ping method, which can give you a little more information on why the ping has failed.

I'm not clear why 4b is failing to fail for you. As long as RaiseError is true, the call to selectrow_array will throw an exception if the server is gone. Perhaps the restart did not restart? Can we see the test output?

**Comment by dr-kd** [comment]
Don't worry about the weird test case.  With the vital info that  ping is not expected to raise an exception I've fixed the test case:

```diff
-eval { $dbh->ping };
-ok $@, '2. $dbh->ping should raise exception if connection is stale, no?';
+my $connected = 0;
+eval { $connected = $dbh->ping };
+ok ! $connected, 'ping failed';
+ok ! $@, '2. $dbh->ping should not raise exception even if connection is stale';

```

I had a look to see if the docs could be improved.  Best I could manage is:

```diff
diff --git a/Pg.pm b/Pg.pm
index ea3237f..ee8ef2f 100644
--- a/Pg.pm
+++ b/Pg.pm
@@ -2813 +2813 @@ Additional information on why a handle is not valid can be obtained by using the
-L</pg_ping> method.
+L</pg_ping> method. <ping> is expected to never raise an exception and always return a value.
```

Reason for reporting the bug in the first place was due to some [historical oracle based fear and confusion](https://metacpan.org/dist/DBIx-Connector/source/lib/DBIx/Connector/Driver/Oracle.pm#L10).

**Comment by turnstep** [comment]
Thanks - made a tweak in 4d52079dd4d1584c60e746774f5cb2391b27eb50

**Comment by turnstep** [comment]
Closed in version 3.18.0
